daemon/daemon_main.go

```diff
@@ -286,21 +287,36 @@ func checkMinRequirements() {
 	if err := os.MkdirAll(globalsDir, defaults.StateDirRights); err != nil {
 		log.WithError(err).WithField(logfields.Path, globalsDir).Fatal("Could not create runtime directory")
 	}
+
+	// Old Cilium BPF feature probes.
+	// TODO(mrostecki): Remove them completely and move missing checks to
+	// bpftool.
 	if err := os.Chdir(option.Config.LibDir); err != nil {
 		log.WithError(err).WithField(logfields.Path, option.Config.LibDir).Fatal("Could not change to runtime directory")
 	}
 	probeScript := filepath.Join(option.Config.BpfDir, "run_probes.sh")
 	if err := exec.Command(probeScript, option.Config.BpfDir, option.Config.StateDir).Run(); err != nil {
 		log.WithError(err).Fatal("BPF Verifier: NOT OK. Unable to run checker for bpf_features")
 	}
-	featuresFilePath := filepath.Join(globalsDir, "bpf_features.h")
+	featuresFilePath := filepath.Join(globalsDir, "bpf_features_legacy.h")
 	if _, err := os.Stat(featuresFilePath); os.IsNotExist(err) {
-		log.WithError(err).WithField(logfields.Path, globalsDir).Fatal("BPF Verifier: NOT OK. Unable to read bpf_features.h")
+		log.WithError(err).WithField(logfields.Path, globalsDir).Fatal("BPF Verifier: NOT OK. Unable to read bpf_features_legacy.h")
 	}
-
 	checkBPFLogs("bpf_requirements", true)
 	checkBPFLogs("bpf_features", false)
-	bpf.ReadFeatureProbes(featuresFilePath)
+
+	// bpftool checks
+	probeManager, err := probes.NewProbeManager()
+	if err != nil {
+		log.WithError(err).Fatal("BPF check: NOT OK.")
+	}
+	if err := probeManager.SystemConfigProbes(); err != nil {
+		log.WithError(err).Fatal("BPF check: NOT OK.")
```

Comment:
I'm pretty sure this will fail in many environments, eg I have an Ubuntu environment happily running current Cilium and here's my `bpftool -j feature` output:

```
$ k -n kube-system exec $(get_cilium_pod) -- bpftool -j feature | jq '.system_config'
{
  "unprivileged_bpf_disabled": 1,
  "bpf_jit_enable": 1,
  "bpf_jit_harden": 0,
  "bpf_jit_kallsyms": 0,
  "bpf_jit_limit": 264241152,
  "CONFIG_BPF": null,
  "CONFIG_BPF_SYSCALL": null,
  "CONFIG_HAVE_EBPF_JIT": null,
  "CONFIG_BPF_JIT": null,
  "CONFIG_BPF_JIT_ALWAYS_ON": null,
  "CONFIG_CGROUPS": null,
  "CONFIG_CGROUP_BPF": null,
  "CONFIG_CGROUP_NET_CLASSID": null,
  "CONFIG_SOCK_CGROUP_DATA": null,
  "CONFIG_BPF_EVENTS": null,
  "CONFIG_KPROBE_EVENTS": null,
  "CONFIG_UPROBE_EVENTS": null,
  "CONFIG_TRACING": null,
  "CONFIG_FTRACE_SYSCALLS": null,
  "CONFIG_FUNCTION_ERROR_INJECTION": null,
  "CONFIG_BPF_KPROBE_OVERRIDE": null,
  "CONFIG_NET": null,
  "CONFIG_XDP_SOCKETS": null,
  "CONFIG_LWTUNNEL_BPF": null,
  "CONFIG_NET_ACT_BPF": null,
  "CONFIG_NET_CLS_BPF": null,
  "CONFIG_NET_CLS_ACT": null,
  "CONFIG_NET_SCH_INGRESS": null,
  "CONFIG_XFRM": null,
  "CONFIG_IP_ROUTE_CLASSID": null,
  "CONFIG_IPV6_SEG6_BPF": null,
  "CONFIG_BPF_LIRC_MODE2": null,
  "CONFIG_BPF_STREAM_PARSER": null,
  "CONFIG_NETFILTER_XT_MATCH_BPF": null,
  "CONFIG_BPFILTER": null,
  "CONFIG_BPFILTER_UMH": null,
  "CONFIG_TEST_BPF": null
}
```

Looks like in this case, we just can't detect the kernel config, so we silently allow Cilium to continue running.

Reply:
OK. Although it's very sad and weird that some distros don't provide kernel config. Are you sure that there is no workaround for Ubuntu.

Yeah, sorry for being focused only on RPM distros. ;)

Reply:
For Ubuntu specifically, I can see a `/boot/config-`uname -r` file which apparently is not picked up / parsed by bpftool.

However, I don't think it's just Ubuntu, AFAIK CoreOS hides the config completely; I wouldn't be surprised if some of the cloud distros skip it as well. I'm all for encouraging distros to provide these configs, but unfortunately I don't think that there's a single isolated distro we can just pressure to fix this, there's a bunch of them.